pkg/builder.go

```diff
@@ -0,0 +1,645 @@
+		return StateNode{}, nil, err
+	}
+	return StateNode{
+		NodeType:  ty,
+		Path:      nodePath,
+		NodeValue: node,
+	}, nodeElements, nil
+}
+
+func (sdb *builder) buildStateTrie(it trie.NodeIterator) ([]StateNode, error) {
+	stateNodes := make([]StateNode, 0)
+	for it.Next(true) {
+		// skip value nodes
+		if it.Leaf() || bytes.Equal(nullHashBytes, it.Hash().Bytes()) {
+			continue
+		}
+		node, nodeElements, err := resolveNode(it, sdb.stateCache.TrieDB())
+		if err != nil {
+			return nil, err
+		}
+		switch node.NodeType {
+		case Leaf:
+			var account state.Account
+			if err := rlp.DecodeBytes(nodeElements[1].([]byte), &account); err != nil {
+				return nil, fmt.Errorf("error decoding account for leaf node at path %x nerror: %v", node.Path, err)
+			}
+			partialPath := trie.CompactToHex(nodeElements[0].([]byte))
+			valueNodePath := append(node.Path, partialPath...)
+			encodedPath := trie.HexToCompact(valueNodePath)
+			leafKey := encodedPath[1:]
+			storageNodes, err := sdb.buildStorageNodesEventual(account.Root, nil, true)
+			if err != nil {
+				return nil, fmt.Errorf("failed building eventual storage diffs for account %+v\r\nerror: %v", account, err)
+			}
+			stateNodes = append(stateNodes, StateNode{
+				NodeType:     node.NodeType,
+				Path:         node.Path,
+				LeafKey:      leafKey,
+				NodeValue:    node.NodeValue,
+				StorageNodes: storageNodes,
+			})
+		case Extension, Branch:
+			stateNodes = append(stateNodes, StateNode{
+				NodeType:  node.NodeType,
+				Path:      node.Path,
+				NodeValue: node.NodeValue,
+			})
+		default:
+			return nil, fmt.Errorf("unexpected node type %s", node.NodeType)
+		}
+	}
+	return stateNodes, it.Error()
+}
+
+// BuildStateDiff builds a statediff object from two blocks and the provided parameters
+func (sdb *builder) BuildStateDiffObject(args Args, params Params) (StateObject, error) {
+	if len(params.WatchedAddresses) > 0 {
+		// if we are watching only specific accounts then we are only diffing leaf nodes
+		log.Info("Ignoring intermediate state nodes because WatchedAddresses was passed")
+		params.IntermediateStateNodes = false
+	}
+
+	// Load tries for old and new states
+	oldTrie, err := sdb.stateCache.OpenTrie(args.OldStateRoot)
+	if err != nil {
+		return StateObject{}, fmt.Errorf("error creating trie for old state root: %v", err)
+	}
+	newTrie, err := sdb.stateCache.OpenTrie(args.NewStateRoot)
+	if err != nil {
+		return StateObject{}, fmt.Errorf("error creating trie for new state root: %v", err)
+	}
+	nWorkers := params.Workers
+	if nWorkers == 0 {
+		nWorkers = 1
+	}
+
+	// Split old and new tries into corresponding subtrie iterators
+	oldIterFac := iter.NewSubtrieIteratorFactory(oldTrie, nWorkers)
+	newIterFac := iter.NewSubtrieIteratorFactory(newTrie, nWorkers)
+	iterChan := make(chan []iterPair, nWorkers)
+
+	// Create iterators ahead of time to avoid race condition in state.Trie access
+	for i := uint(0); i < nWorkers; i++ {
+		// two state iterations per diff build
```

Comment:
Could you explain the reasoning behind needing the two iterations? I can by playing with the unit tests that this is necessary, but I'm having a hard time wrapping my head around why (this stuff is freaking tricky and I haven't thought about it in a hot minute).

Reply:
Yeah, the first pass identifies created/updated nodes, second pass is deleted/updated. They both use the difference iterator with arguments reversed - i.e. first pass wants nodes in B but not A (B \ A), second wants A \ B.